### PR TITLE
resource: fix precision loss and division-by-zero in resource conversion

### PR DESCRIPTION
**Pull request by antcybersec**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
The `convertResource*ToString` helpers in `pkg/api/v1/resource`, `staging/src/k8s.io/kubectl/pkg/util/resource`, and `staging/src/k8s.io/kubectl/pkg/cmd/set/env` relied on `float64` division (e.g. `float64(cpu.MilliValue()) / float64(divisor.MilliValue())`), which led to two problems:
1. Division-by-zero / overflow when `divisor.MilliValue()` or `divisor.Value()` evaluated to 0 (for instance, when using sub-millicores like `1n` or `1u` as divisors).
2. Floating-point precision loss during ceiling division (e.g., `0.001 / 0.000001` yielding `1001` instead of `1000` due to float64 binary representation).

This PR refactors those helpers to use exact `math/big.Int` decimal scale math, computing $\text{ceil}(N / D) = (N + D - 1) / D$.

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:
Added test cases in `pkg/api/v1/resource/helpers_test.go` covering sub-milli divisors (`1n`, `1u`) and zero divisors.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by antcybersec** [comment]
@ardaguclu Hi! Gentle ping for review and `/ok-to-test` when you get a chance. Thanks!

**Comment by antcybersec** [comment]
For reviewers' visibility: #142265 was opened later (09-21) and covers the same four files with the same fix. This PR was opened 09-18. Flagged there as well so the two can be consolidated.

**Comment by antcybersec** [comment]
@thc1006 Hi! Gentle ping for `/ok-to-test` on this PR (opened Sep 18, prior to duplicate #142265). Thanks!

**Comment by thc1006** [comment]
I just posted `/ok-to-test`, but I’m curious: if you knew this PR come up duplicate PRs and knew where the issue you’re addressing originated, why didn’t you metion the origin issue at first?

just curious just curious not say ur wrong.

**Comment by antcybersec** [comment]
@thc1006 Thanks for adding `/ok-to-test` and for curating the tracker!

That was an honest oversight on my end when filling out the initial PR description template—I was focused on getting the test cases and `math/big.Int` ceiling math refactor ready and omitted `Part of #141166` under the issue section. I've updated the description to link `Part of #141166`. Appreciate you pointing it out and keeping everything organized!

File: pkg/api/v1/resource/helpers.go [hunk 1]
```diff
@@ -142,32 +141,69 @@ func ExtractContainerResourceValue(fs *v1.ResourceFieldSelector, container *v1.C
...
+	qBig := new(big.Int).Set(qDec.UnscaledBig())
+	divBig := new(big.Int).Set(divDec.UnscaledBig())
+
+	qScale := int(qDec.Scale())
+	divScale := int(divDec.Scale())
+
+	sDiff := divScale - qScale
+
+	if sDiff > 0 {
+		exp := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(sDiff)), nil)
```
Comment (thc1006) [inline pkg/api/v1/resource/helpers.go:185]: This builds 10^(scale difference) with no bound. A memory value written as 1e1000000 with divisor 1 already comes out as a million digits here.

**Review by thc1006 — COMMENTED** [review]
The bound reads right now. On `7017077` a memory value of `1e1000000` over `1` comes back saturated in a few microseconds, and `1` over `1e2147483647` gives `1`. One edge on the saturation itself: `9223372036854775808` over `1` still comes back as itself and `10E` as `10000000000000000000`, only `1e20` and up land on `MaxInt64`, so the int64 the comment promises starts one order of magnitude late. Master returns `-9223372036854775808` for all of those. Still leaving the pick between this and #142265 to @jpbetz.

File: pkg/api/v1/resource/helpers.go [hunk 2]
```diff
@@ -142,32 +143,102 @@ func ExtractContainerResourceValue(fs *v1.ResourceFieldSelector, container *v1.C
...
+	if divBig.Sign() <= 0 {
+		return "0", nil
+	}
+
+	// ceil(N / D) = (N + D - 1) / D
+	tmp := new(big.Int).Add(qBig, divBig)
+	tmp.Sub(tmp, big.NewInt(1))
+	res := new(big.Int).Quo(tmp, divBig)
+
+	return res.String(), nil
```
Comment (thc1006) [inline pkg/api/v1/resource/helpers.go:201]: The result isn't range-checked after the ceiling, so the output isn't monotonic either: `7e19` over `1` comes back as `70000000000000000000` and `8e19` over `1` as `9223372036854775807`. `resource.go:127` and `env_resolve.go:275` in kubectl return the same way.

File: pkg/api/v1/resource/helpers.go [hunk 3]
```diff
@@ -142,32 +143,102 @@ func ExtractContainerResourceValue(fs *v1.ResourceFieldSelector, container *v1.C
...
+		return "0", nil
+	}
+
+	qDec := q.AsDec()
+	divDec := divisor.AsDec()
+
+	qBig := new(big.Int).Set(qDec.UnscaledBig())
+	divBig := new(big.Int).Set(divDec.UnscaledBig())
+
+	qScale := int(qDec.Scale())
```
Comment (thc1006) [inline pkg/api/v1/resource/helpers.go:162]: These are `int`, and kubectl still builds for `linux/386`, `linux/arm` and `windows/386`. There a parsed `8e2147483647` over `1` has scale `-2147483647`, both differences wrap to `MinInt32`, neither shortcut fires, and the alignment goes for `10^2147483647`. Widening to `int64` before the subtraction avoids it.

File: pkg/api/v1/resource/helpers.go [hunk 4]
```diff
@@ -142,32 +143,102 @@ func ExtractContainerResourceValue(fs *v1.ResourceFieldSelector, container *v1.C
...
+}
+
+// maxResultDigits is the number of decimal digits math.MaxInt64 has. A ratio
+// wider than this cannot be expressed as an int64, so it saturates.
+const maxResultDigits = 19
+
+// decimalExponent returns the approximate power of ten of unscaled/10^scale.
+// It reads the bit length rather than the digits so that measuring a value
+// like 1e1000000 stays cheap.
+func decimalExponent(unscaled *big.Int, scale int) int {
```
Comment (thc1006) [inline pkg/api/v1/resource/helpers.go:211]: The estimate reads the representation, not the value. A divisor equal to `1` built with `NewDecimalQuantity` as unscaled `10^10000` at scale `10000` makes `2` over it come back as `1`. Parsed quantities never carry a padded coefficient like that, so this is only about the helper's contract.

File: pkg/api/v1/resource/helpers.go (hunk 2 above)
Comment (antcybersec) [reply pkg/api/v1/resource/helpers.go]: Range-checked the quotient with `res.IsInt64()` in ab5b427a1df to saturate at `math.MaxInt64` across all three files.

File: pkg/api/v1/resource/helpers.go (hunk 3 above)
Comment (antcybersec) [reply pkg/api/v1/resource/helpers.go]: Widened scales and scale differences to `int64` in ab5b427a1df to prevent `MinInt32` wrapping on 32-bit architectures.

File: pkg/api/v1/resource/helpers.go (hunk 4 above)
Comment (antcybersec) [reply pkg/api/v1/resource/helpers.go]: Replaced point estimation with `decimalExponentBounds` in ab5b427a1df so padded coefficients cannot trigger premature shortcuts.

**Comment by antcybersec** [comment]
@thc1006 @jpbetz Updated with commit ab5b427a1df addressing all review comments (range checks, int64 scale widening, and decimalExponentBounds)—PTAL when you get a chance! Thanks!

**Comment by antcybersec** [comment]
@jpbetz Gentle ping whenever you have a moment to weigh in on consolidating between this PR (opened Sep 18, fully updated with edge-case tests per @thc1006's review) and duplicate #142265. Thanks!
